### ISSUE TITLE
[Improvement]: Refactor `BUnionType`'s  `getMemberTypes()` usages

### ISSUE THREAD

**Issue by lochana-chathura** [ballerina-platform/ballerina-lang#43344]
### Description

$subject. We need to get rid of `getMemberTypes()` and `getOriginalMemberTypes()` usages in the current compiler, in order to do #43343.

This issue depends on #43126 which is in PR sent state now.

**Comment by lochana-chathura** [comment]
- Found a bug in `isSimpleNode()` where `AllOrNothingSubtype` instance checks are carried out instead of `BddAllOrNothing` instance checks. This resulted in not giving correct values for `Core.mappingAtomicType()`.
- New API method `Optional<List<MappingAtomicType>> mappingAtomicTypesInUnion(Context, SemType)` introduced to implement `isLaxType()` check using semtypes.

Fixed with [91ce09e](https://github.com/ballerina-platform/ballerina-lang/pull/43341/commits/91ce09e39f908d4b44140e1970ea8b1a93f71e2b)

**Comment by lochana-chathura** [comment]
When converting isSameType() usages to semtypes, here at
https://github.com/ballerina-platform/ballerina-lang/blob/d93ec613f5e22f67a868593a66493119d860f201/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ServiceDesugar.java#L271

the same BObjectType instance with one inside a BUnionType wrapper, evaluate not to be the same type, in the current compiler. 

<img width="403" alt="Screenshot 2024-10-02 at 07 50 51" src="https://github.com/user-attachments/assets/09862213-b3bc-4bbf-a6d7-12303e705dd5">

We are comparing BUnionType 5262 to BObjectType 5231

With semTypes, they both identified to be the same type, causing NPE at later stage

**Comment by lochana-chathura** [comment]
When rewriting `isDistinctAllowedOnType()` using semtypes, it was found that the same method is called twice for an `BObjectType` causing the wrong resolution of distinct type IDs for the semtype. When calling [1], `BTypeIdSet` info is not available, resulting in the wrong semtype resolution.

Two places:
[1] https://github.com/ballerina-platform/ballerina-lang/blob/fe0509bed414ca8e1014f0bbb28e1f7a492de09d/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeResolver.java#L1098

[2] https://github.com/ballerina-platform/ballerina-lang/blob/fe0509bed414ca8e1014f0bbb28e1f7a492de09d/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeResolver.java#L729

Removing [1] fixed the issue.

**Comment by lochana-chathura** [comment]
Found another issue related to `isDistinctAllowedOnType()`. When calling it at Symbol Resolver, the BObject type's symbol does not have the `attachedFuncs` values updated, causing a wrong resolution of semtype for the object type. 

```bal
type Obj0 distinct object {
    function apply() returns int;
};

type Obj2 object {
    function apply() returns int;
};

function foo(Obj0 x) {
    Obj2 y = x; // Should have no error
}
```

**Comment by lochana-chathura** [comment]
After rewriting `isAllReadonlyTypes()` using semtypes at,
https://github.com/ballerina-platform/ballerina-lang/blob/fe0509bed414ca8e1014f0bbb28e1f7a492de09d/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java#L3588

the following test that used compile will now give an error,
```bal
class P1 {
    Q1 q;
    string p1;

    function init() {
        self.q = new;
        self.p1 = "P1";
    }
}

class Q1 {
    P1 p;
    string q1;

    function init() {
        self.p = new;
        self.q1 = "Q1";
    }
}

function testArraysOfCyclicDependentTypes3() returns P1[] {
    P1[] arr = [];
    arr[3] = new; // ERROR [test_parser.bal:(23:5,23:11)] cannot update 'readonly' value of type 'P1[]'
    return arr;
}

function testArraysOfCyclicDependentTypes4() returns Q1[] {
    Q1[] arr = [];
    arr[3] = new; // ERROR [test_parser.bal:(29:5,29:11)] cannot update 'readonly' value of type 'Q1[]'
    return arr;
}
```

This is because in semtype P1 and P2 are considered to have an empty value space making them equivalent to `never`

**Comment by lochana-chathura** [comment]
const int[2]|[int, int] ERR_TUPLE5 = []; // ambiguous types

With the use of `isSameType()` semtype based API, the above will no longer give the ambiguous type error.

**Comment by lochana-chathura** [comment]
Sample:
```bal
function testBitWiseOperationsForNullable() {
    byte? x = 251;
    int:Signed8? y = -123;

    var ans7 = x >> y;
}
```
Replacing the following `isSameType()` with semtype-based API would crash as `function (int:Unsigned8?,int:Signed8?) returns (int:Unsigned8?)` considered to be resolved operator symbol rather than `function (byte?,int:Signed8?) returns (byte?)`.

https://github.com/ballerina-platform/ballerina-lang/blob/b67b420361dc9db1d8860c8a17b3efadebd60aa0/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java#L2205

BIR diff:
<img width="1126" alt="Screenshot 2024-10-08 at 15 29 03" src="https://github.com/user-attachments/assets/fd8af085-bd4c-4679-bdb2-8fcf676ae426">

<details>
  <summary><b>crash log</b>  (<i>click to expand</i>)</summary>

```
org.ballerinalang.compiler.BLangCompilerException: error while generating method 'testBitWiseOperationsForNullable' in class 'test_parser'
	at org.wso2.ballerinalang.compiler.bir.codegen.JvmCodeGenUtil.visitMaxStackForMethod(JvmCodeGenUtil.java:760)
	at org.wso2.ballerinalang.compiler.bir.codegen.methodgen.MethodGen.genJMethodForBFunc(MethodGen.java:390)
	at org.wso2.ballerinalang.compiler.bir.codegen.methodgen.MethodGen.generateMethod(MethodGen.java:186)
	at org.wso2.ballerinalang.compiler.bir.codegen.JvmPackageGen.lambda$generateModuleClasses$0(JvmPackageGen.java:420)
	at java.base/java.util.HashMap.forEach(HashMap.java:1421)
	at org.wso2.ballerinalang.compiler.bir.codegen.JvmPackageGen.generateModuleClasses(JvmPackageGen.java:377)
	at org.wso2.ballerinalang.compiler.bir.codegen.JvmPackageGen.generate(JvmPackageGen.java:765)
	at org.wso2.ballerinalang.compiler.bir.codegen.CodeGenerator.generate(CodeGenerator.java:86)
	at org.wso2.ballerinalang.compiler.bir.codegen.CodeGenerator.generate(CodeGenerator.java:63)
	at io.ballerina.projects.JBallerinaBackend.performCodeGen(JBallerinaBackend.java:377)
	at io.ballerina.projects.ModuleContext.generateCodeInternal(ModuleContext.java:458)
	at io.ballerina.projects.ModuleCompilationState$4.generatePlatformSpecificCode(ModuleCompilationState.java:132)
	at io.ballerina.projects.ModuleContext.generatePlatformSpecificCode(ModuleContext.java:358)
	at io.ballerina.projects.JBallerinaBackend.performCodeGen(JBallerinaBackend.java:178)
	at io.ballerina.projects.JBallerinaBackend.<init>(JBallerinaBackend.java:144)
```
</details>

**Comment by lochana-chathura** [comment]
When replacing the `types.isSameBIRShape()` call in `BIRPackageSymbolEnter` with semtype based `isSameType()`, encountered with the following crash.

<details>
  <summary><b>stacktrace</b>  (<i>click to expand</i>)</summary>

```
ballerina-new-parser-test-suite > jballerina-test > org.ballerinalang.test.bala.types.TypeNameWithSpecialCharsBalaTest > testTypeNameWithSpecialChars FAILED
    org.ballerinalang.compiler.BLangCompilerException: failed to load the module 'specialchars/pkg:0.0.0' from its BIR due to: Cannot invoke "org.wso2.ballerinalang.compiler.semantics.model.types.BType.semType()" because "this.referredType" is null
        at app//org.wso2.ballerinalang.compiler.BIRPackageSymbolEnter.definePackage(BIRPackageSymbolEnter.java:239)
        at app//org.wso2.ballerinalang.compiler.BIRPackageSymbolEnter.definePackage(BIRPackageSymbolEnter.java:215)
        at app//io.ballerina.projects.ModuleContext.loadPackageSymbolInternal(ModuleContext.java:528)
        at app//io.ballerina.projects.ModuleCompilationState$6.compile(ModuleCompilationState.java:178)
        at app//io.ballerina.projects.ModuleContext.compile(ModuleContext.java:354)
        at app//io.ballerina.projects.PackageCompilation.compileModulesInternal(PackageCompilation.java:211)
        at app//io.ballerina.projects.PackageCompilation.compileModules(PackageCompilation.java:195)
        at app//io.ballerina.projects.PackageCompilation.compile(PackageCompilation.java:102)
        at app//io.ballerina.projects.PackageCompilation.from(PackageCompilation.java:97)
        at app//io.ballerina.projects.PackageContext.getPackageCompilation(PackageContext.java:243)
        at app//io.ballerina.projects.Package.getCompilation(Package.java:154)
        at app//org.ballerinalang.test.BCompileUtil.jBallerinaBackend(BCompileUtil.java:224)
        at app//org.ballerinalang.test.BCompileUtil.compile(BCompileUtil.java:99)
        at app//org.ballerinalang.test.bala.types.TypeNameWithSpecialCharsBalaTest.testTypeNameWithSpecialChars(TypeNameWithSpecialCharsBalaTest.java:43)

        Caused by:
        java.lang.NullPointerException: Cannot invoke "org.wso2.ballerinalang.compiler.semantics.model.types.BType.semType()" because "this.referredType" is null
            at org.wso2.ballerinalang.compiler.semantics.model.types.BTypeReferenceType.semType(BTypeReferenceType.java:44)
            at org.wso2.ballerinalang.compiler.semantics.model.types.BRecordType.semType(BRecordType.java:171)
            at org.wso2.ballerinalang.compiler.semantics.analyzer.SemTypeHelper.semType(SemTypeHelper.java:165)
            at org.wso2.ballerinalang.compiler.semantics.analyzer.SemTypeHelper.semType(SemTypeHelper.java:109)
            at org.wso2.ballerinalang.compiler.semantics.analyzer.Types.isSameTypeWithTags(Types.java:406)
            at org.wso2.ballerinalang.compiler.semantics.analyzer.Types.isSameBIRShape(Types.java:2498)
            at org.wso2.ballerinalang.compiler.BIRPackageSymbolEnter.getType(BIRPackageSymbolEnter.java:2166)
            at org.wso2.ballerinalang.compiler.BIRPackageSymbolEnter$BIRTypeReader.readType(BIRPackageSymbolEnter.java:1364)
            at org.wso2.ballerinalang.compiler.BIRPackageSymbolEnter.readBType(BIRPackageSymbolEnter.java:715)
            at org.wso2.ballerinalang.compiler.BIRPackageSymbolEnter$BIRTypeReader.readTypeFromCp(BIRPackageSymbolEnter.java:1185)
            at org.wso2.ballerinalang.compiler.BIRPackageSymbolEnter$BIRTypeReader.readType(BIRPackageSymbolEnter.java:1551)
            at org.wso2.ballerinalang.compiler.BIRPackageSymbolEnter.readBType(BIRPackageSymbolEnter.java:715)
            at org.wso2.ballerinalang.compiler.BIRPackageSymbolEnter$BIRTypeReader.readTypeFromCp(BIRPackageSymbolEnter.java:1185)
            at org.wso2.ballerinalang.compiler.BIRPackageSymbolEnter$BIRTypeReader.readType(BIRPackageSymbolEnter.java:1385)
```
</details> 


```bal
public type Bdd Node|boolean;

public type Node readonly & record {|
    int atom;
    Bdd left;
    Bdd middle;
    Bdd right;
|};
```

The reason is in the `Node` record type, the referred types for `left`, `middle` and `right` are `null` when calling at `BIRPackageSymbolEnter`. However, via the previous API since we don't access the type(referred type's symbol is used instead) it was not a problem.

Failing test: `TypeNameWithSpecialCharsBalaTest`

**Comment by lochana-chathura** [comment]
>The reason is in the Node record type, the referred types for left, middle and right are null when calling at BIRPackageSymbolEnter. However, via the previous API since we don't access the type(referred type's symbol is used instead) it was not a problem.

The reason for `null` was that the type was still resolving. For `BReferenceType` we update the referredType after creating the instance. It turned out `isSameBIRShape()` was a temporary workaround(See: https://github.com/ballerina-platform/ballerina-lang/issues/35872#issuecomment-1108144686) and the original issue has been fixed with PR #36168.

Hence, `isSameBIRShape()` can be removed.

**Comment by lochana-chathura** [comment]
- `BSameTypeVisitor` and `BIRSameShapeVisitor` visitors have now been removed.  The dormer was replaced with the `isSameType()` API from semtype module.
- `BOrderedTypeVisitor` has been removed. Used `comparable()` API instead, from nBallerina. For that needed to port extra few APIs from there. Additionally, in the nBallerina API, there was a missing part because of that the following is considered comparable in nBallerina. Need to check if the accumulated types belong to same ordered type if the lengths are different.

```bal
type X [(), (), (), (), int, (), (), string];
type Y [5.0];
```

**Comment by lochana-chathura** [comment]
Found an inconsistency in current interop method return type validation.
e.g. 
ballerina functions:
```bal
function getNilAsReadOnly() returns error? = @java:Method { // <---------------[1]
    'class: "org/ballerinalang/nativeimpl/jvm/tests/StaticMethods"
} external;

function getNilAsReadOnly() returns readonly = @java:Method { // <---------------[2]
    'class: "org/ballerinalang/nativeimpl/jvm/tests/StaticMethods"
} external;

function getBooleanAsReadOnly() returns readonly = @java:Method { <---------------[3]
    'class: "org/ballerinalang/nativeimpl/jvm/tests/StaticMethods"
} external;
````
java methods:
```java
    public static void getNilAsReadOnly() {
    }

    public static boolean getBooleanAsReadOnly() {
        return true;
    }
```

For [1] a compile error is given because `error` is present in the return type, however java method does give exception or anything.
For [2] and [3] no compiler error. Although it is not explicitly declared `readonly` does have the error within. 

Inconsistency is fixed with [e70628b](https://github.com/ballerina-platform/ballerina-lang/pull/43341/commits/e70628b5e491ebd3c2d14ebc5144e761794603e3). In the new behavior all 3 cases are allowed. No compile error.

**Comment by lochana-chathura** [comment]
The following bal code that was used to compile now gives compile-time errors commented down. 
```bal
import ballerina/test;

type IntOne 1;
type FloatOne 1.0;
type IntTwo 2;
type FloatTwo 2f;

function checkFiniteTypeEquality() {
    IntOne intOne_1 = 1;
    IntOne intOne_2 = 1;
    IntTwo intTwo = 2;
    FloatOne floatOne = 1f;
    FloatTwo floatTwo = 2.0;

    test:assertTrue((intOne_1 == intOne_2) && !(intOne_1 != intOne_2));
    test:assertTrue((floatOne != floatTwo) && !(floatOne == floatTwo)); // Line 16
    test:assertFalse((intOne_1 == intTwo) && !(intOne_1 != intTwo)); // Line 17
}

type Array ["array", 1];
type Mapping ["mapping", 2];

function checkTupleEqualityNegative() {
    Array a = ["array", 1];
    Mapping b = ["mapping", 2];
    test:assertFalse(a == b); // Line 26
}

//ERROR [test_parser.bal:(16:22,16:42)] operator '!=' not defined for 'FloatOne' and 'FloatTwo'
//ERROR [test_parser.bal:(16:49,16:69)] operator '==' not defined for 'FloatOne' and 'FloatTwo'
//ERROR [test_parser.bal:(17:23,17:41)] operator '==' not defined for 'IntOne' and 'IntTwo'
//ERROR [test_parser.bal:(17:48,17:66)] operator '!=' not defined for 'IntOne' and 'IntTwo'
//ERROR [test_parser.bal:(26:22,26:28)] operator '==' not defined for 'Array' and 'Mapping'
```

According to [equality-expr](https://ballerina.io/spec/lang/master/#equality-expr) section in spec,
> An equality-expr tests whether two values are equal. For all four operators, it is a compile time error if the intersection of the static types of the operands is empty. 

So the new behavior is correct. We need to improve the compile time error message.

**Comment by lochana-chathura** [comment]
In case of inferred isolation, the isolated tag is added at `IsolationAnalyzer`. This will mutate the semtype.  Therefore, [9af49c6](https://github.com/ballerina-platform/ballerina-lang/pull/43341/commits/9af49c622ca752aae73d6f94e68abe3ba26bdf94) was added to reset semtype in case of a flag mutation.

With that fix `EscapedIdentifiersValidationTest` failure was resolved. However, the fix uncovered a bug in the existing implementation. The affected test case is `FunctionTests.testFunctionalProgramming`. 

In the previous `BSameTypeVisitor`, two functions were assumed to be the same if the source has the `isolated` flag and the target does not have the `isolated` flag. Since now they are not evaluated to the same, a new ImplicitCastExpr is added here https://github.com/ballerina-platform/ballerina-lang/blob/8ef9be112ff1bb02d8f647fb2aadced794417941/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java#L1535. In the new cast we ignore the `BReferenceType` wrapper as per the following logic.
https://github.com/ballerina-platform/ballerina-lang/blob/8ef9be112ff1bb02d8f647fb2aadced794417941/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java#L1531-L1533 

Therefore, the resultant runtime type will not have the reference type name propagated.
Old type string: IntFilter
New type string: isolated function (int) returns (boolean)

This caused the `FunctionTests.testFunctionalProgramming` failure.

**Comment by lochana-chathura** [comment]
After rewriting the `isTypeCastable()` method with seam types, the following samples used give compile-time errors, now compile with no errors.
1. BasicTupleTest.java
```bal
function testTupleToJSONAssignmentNegative() {
    xml A = xml `xml string`;
    [string, int|xml, string...] C = ["text1", 1, A.toString()];
    json jsonTest = <json[]>C; // No compile-time error for the cast.
}
```
2. TableCastTest.java
```bal
type Teacher record {
    readonly string name;
    readonly int age;
    string school;
};

type Customer record {
    readonly int id;
    readonly string name;
    string lname;
};

type CustomerTable table<Customer|Teacher>;

CustomerTable tab3 = table key(name) [
    {id: 13, name: "Foo", lname: "QWER"},
    {id: 13, name: "Bar", lname: "UYOR"}
];

type Customer2 record {|
    int id;
    string name;
    string lname;
|};

type CustomerEmptyKeyedTbl table<Customer2> key();

function testCastingWithEmptyKeyedKeylessTbl() {
    CustomerEmptyKeyedTbl tbl1 = <CustomerEmptyKeyedTbl>tab3; // No compile-time error for the cast.

    table<record {
        int id;
        string name;
        string lname;
    }> key() tbl = <table<record {
        int id;
        string name;
        string lname;
    }> key()>tab3; // No compile-time error for the cast.
}
```

In both cases, non-empty intersections exist.  Therefore new behavior is acceptable. Need to make sure these casts are not crashing at runtime.

**Comment by lochana-chathura** [comment]
After rewriting `intersectionExists()` method in `CodeAnalyzer.java` using semtype intersection,
The following samples that used to go to runtime, now give errors at compile-time. 
```bal
function testNeverRuntime10() {
    int x = 100;
    boolean b = x is never; // incompatible types: 'int' will not be matched to 'never'
    assertEquality(false, b);
}

type Record record {|
    int i;
    never[] j;
|};


function testNeverRuntime11() {
    Record x = { i: 1, j: [] };
    boolean b = x is never; // incompatible types: 'Record' will not be matched to 'never'
    assertEquality(false, b);
}
```

```bal
function testNeverFieldTypeCheck() {
    record {int x;} r2 = {x: 2, "color": "blue"};
    assertEquality(false, r2 is record {never x?;}); // incompatible types: 'record {| int x; anydata...; |}' will not be matched to 'record {| never x?; anydata...; |}'

    record {never? x;} r3 = {x: (), "color": "blue"};
    assertEquality(false, r3 is record {never x?;}); // incompatible types: 'record {| never? x; anydata...; |}' will not be matched to 'record {| never x?; anydata...; |}'

    record {int? x;} r4 = {x: 2, "color": "blue"};
    assertEquality(false, r4 is record {never x?;}); // incompatible types: 'record {| int? x; anydata...; |}' will not be matched to 'record {| never x?; anydata...; |}'

    record {int x;} & readonly v3 = {x: 2, "color": "blue"};
    assertEquality(false, v3 is record {never x?;}); // incompatible types: '(record {| int x; anydata...; |} & readonly)' will not be matched to 'record {| never x?; anydata...; |}'

    record {never? x;} & readonly v4 = {x: (), "color": "blue"};
    assertEquality(false, v4 is record {never x?;}); // incompatible types: '(record {| never? x; anydata...; |} & readonly)' will not be matched to 'record {| never x?; anydata...; |}'
}
```

**Comment by lochana-chathura** [comment]
Closing with #43341
